roadmap/implementers-guide/src/runtime/scheduler.md

```diff
@@ -1,3 +1,264 @@
+  /// Returns an error, if index is out of bounds. Returns `Ok(None)` if no
+  /// furhter assignments are available for that core.
+  pub fn pop_assignment_for_core(&mut self, core_id: CoreId) -> Result<Option<ParaId>>;
+}
+```
+
+On each block, the scheduler will check its claim queue sizes for each core.
+Then it calls `pop_assignment_for_core` on its given `AssignmentProvider` for
+each core where the claim queue size is smaller than desired, until the queue is
+either full or `pop_assignment_for_core` returns `None`.
+
+Whenever a core becomes free, we reduce the corresponding `ClaimQueue` size by
+one.
+
+A single implementation of `AssignmentProvider` needs to be passed to the
+scheduler, internally the given `AssignmentProvider` can delegate to other providers
+handling sub areas, but that is completely transparent to the scheduler. E.g.:
+
+```rust
+struct PolkadotAssignmentProvider {
+  parathread_provider: ParathreadProvider,
+  parachain_provider: ParachainProvider,
+}
+
+impl AssignmentProvider for PolkadotAssignmentProvider {
+  fn pop_assignment_for_core(&mut self, core_id: CoreId) -> Result<Option<ParaId>> {
+     match self.get_sub_index(core_id) {
+       SubIndex::Parathread(thread_id) =>
+         self.parathread_provider.pop_assignment(thread_id),
+       SubIndex::Parachain(chain_id) =>
+         self.parachain_provider.pop_assignment(chain_id),
+     }
+   }
+}
+// What core ranges belong to which provider will be configurable and can change
+// at session boundaries.
+```
+
+The providers need to be able to handle some cores moving faster than others.
+For example, if availability is slower on one core for some reason, then the
+scheduler will call `pop_assignment_for_core` less on that core ... it will
+impose some back pressure. The provider needs to be able to handle this.
+
+Open questions:
+
+With this interface we will charge parathreads fully once
+`pop_assignment_for_core` got called, no matter what happens next. That is even
+on error. This adds to simplicity, great if acceptable ... I think at least for
+a first run it is. This is one of those things that can be improved, once it is
+actually causing problems. -> Especially with async backing it is actually quite
+unlikely that things fail.
+
+
+### Interface from the scheduler to the node size (collators/validators)
+
+We have been a bit fuzzy with the claim queue here, time to make it more
+concrete. The claim queue is used internally by the scheduler to be able to
+satisfy a lookahead on assignments. It could look something like this:
+
+```rust
+/// A claim queue for a single core:
+struct ClaimQueue {
+  /// Actual queue:
+  queue: Vec<ParaId>
+  /// A claim previously in queue directly, that is now pending availability.
+  ///
+  /// It is counting towards the queue size, in the sense that we only request
+  /// another assignment once the core is freed again.
+  pending_availability: PendingAvailabilityClam,
+
+}
+
+struct PendingAvailabilityClaim {
```

Comment:
This needs to include candidate hash, at least if it is intended to replace the `fn availability_cores()` runtime API. The provisioner in asynchronous backing needs to know which candidate it has the bitfields to include, so it can provide a candidate that's a child of it, if necessary.